### ISSUE TITLE
Duplicates found in y_true list

### ISSUE THREAD

**Issue by Shadz13** [massquantity/LibRecommender#38]
**Current Version : 0.4.0**

In function print_metrics_ranking, I have added in the lines:
```
        rec_list = pd.DataFrame.from_dict(y_reco_list, orient="index").reset_index()
        rec_list.columns = ['user', 'rec_1','rec_2','rec_3', 'rec_4','rec_5','rec_6', 'rec_7','rec_8','rec_9', 'rec_10']
        true_list = pd.DataFrame.from_dict(y_true_list, orient="index").reset_index()
        true_list.columns = ['user', 'true_1', 'true_2', 'true_3' ,'true_4']
        
        
        print(true_list) 
        print(rec_list) 
```
I have noticed the following from the printed dataframe:
1.	In y_true_list, there can be more than more item and in one example the item number is repeated. An example output is shown below:
9089:array('I', [7, 7, 7])
9023: array('I', [17, 14])
Is it possible to have more than one item per user and have duplicate items for a specific user?

A full version of the updated function can be found below:
```
def print_metrics_ranking(self,df,metrics, y_prob=None, y_true=None, y_reco_list=None,
                          y_true_list=None, users=None, k=10, train=True):
    #global eval_df
    #global v
    eval_df=pd.DataFrame()
    
    if train:
        for m in metrics:
            if m in ["log_loss", "loss"]:
                log_loss_ = log_loss(y_true, y_prob, eps=1e-7)
                print(f"\t train log_loss: {log_loss_:.4f}")
                eval_df['log_loss']=[log_loss_]      
    else:
        for m in metrics:

            if m in ["log_loss", "loss"]:
                log_loss_ = log_loss(y_true, y_prob, eps=1e-7)
                print(f"\t eval log_loss: {log_loss_:.4f}")
                eval_df['log_loss_']=[log_loss_]
                
            elif m == "balanced_accuracy":
                y_pred = np.round(y_prob)
                accuracy = balanced_accuracy_score(y_true, y_pred)
                print(f"\t eval balanced accuracy: {accuracy:.4f}")
                eval_df['accuracy']=[round(accuracy,4)]
                
                
            elif m == "roc_auc":
                roc_auc = roc_auc_score(y_true, y_prob)
                print(f"\t eval roc_auc: {roc_auc:.4f}")
                eval_df['roc_auc']=[round(roc_auc,4)]

            elif m == "pr_auc":
                precision, recall, _ = precision_recall_curve(y_true, y_prob)
                pr_auc = auc(recall, precision)
                eval_df['pr_auc']=[round(pr_auc,4)]
                
            elif m == "precision":
                precision_all = precision_at_k(y_true_list, y_reco_list,
                                               users, k)
                print(f"\t eval precision@{k}: {precision_all:.4f}")
                eval_df['precision_all']=[round(precision_all,4)]
                
            elif m == "recall":
                recall_all = recall_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval recall@{k}: {recall_all:.4f}")
                eval_df['recall_all']=[round(recall_all,4)]
                
            elif m == "map":
                map_all = map_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval map@{k}: {map_all:.4f}")
                eval_df['map_all']=[round(map_all,4)]
                
            elif m == "ndcg":
                ndcg_all = ndcg_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval ndcg@{k}: {ndcg_all:.4f}")
                eval_df['ndcg_all']=[round(ndcg_all,4)]
                
        #print(y_true_list)
        
        df = pd.concat([df,eval_df],ignore_index=True)
        
        #print(test_results_df)
        rec_list = pd.DataFrame.from_dict(y_reco_list, orient="index").reset_index()
        rec_list.columns = ['user', 'rec_1','rec_2','rec_3', 'rec_4','rec_5','rec_6', 'rec_7','rec_8','rec_9', 'rec_10']
        true_list = pd.DataFrame.from_dict(y_true_list, orient="index").reset_index()
        true_list.columns = ['user', 'true_1', 'true_2', 'true_3' ,'true_4', 'true_5', 'true_6', 'true_7' ,'true_8', 'true_9', 'true_10', 'true_11']
        self.rec_results  = pd.concat([self.rec_results ,rec_list],ignore_index=True)
        self.true_results  = pd.concat([self.true_results ,true_list],ignore_index=True)
        
    return df,self.true_results,self.rec_results

```

**Comment by Shadz13** [comment]
In function print_metrics_ranking, I have added in the lines:
```
        df = pd.concat([df,eval_df],ignore_index=True)
        test_results = pd.concat([test_results,test_results_df],ignore_index=True)
        rec_list = pd.DataFrame.from_dict(y_reco_list, orient="index").reset_index()
        rec_list.columns = ['user', 'rec_1','rec_2','rec_3', 'rec_4','rec_5','rec_6', 'rec_7','rec_8','rec_9', 'rec_10']
        true_list = pd.DataFrame.from_dict(y_true_list, orient="index").reset_index()
        true_list.columns = ['user', 'true_1', 'true_2', 'true_3' ,'true_4']
        
        
        print(true_list) 
        print(rec_list) 
```
I have noticed following from the printed dataframes:
1) The number of rows for the true_list and rec_list are not the same
![image](https://user-images.githubusercontent.com/42578613/105746687-6e6f3d00-5f48-11eb-9355-f36f4b9aedb1.png)
2) In y_true_list there can be more than more item and in one example the item number is repeated
`9089:array('I', [7, 7, 7])`
`9023: array('I', [17, 14])`

**Comment by massquantity** [comment]
Could you provide your whole modified `print_metrics_ranking` function? I couldn't understand the logic without it...

**Comment by Shadz13** [comment]
I am currently using a small data set to evaluate the code and update the script. It contains the following:.
```
self.n_users  10000
self.n_items  42
```
```
def print_metrics_ranking(df,test_counter,train_counter,train_results,test_results,metrics, y_prob=None, y_true=None, y_reco_list=None,
                          y_true_list=None, users=None, k=10, train=True):
    #global eval_df
    #global v
    eval_df=pd.DataFrame()
    train_results_df=pd.DataFrame()
    test_results_df=pd.DataFrame()
    
    if train:
        train_counter=train_counter+1
        for m in metrics:
            if m in ["log_loss", "loss"]:
                log_loss_ = log_loss(y_true, y_prob, eps=1e-7)
                print(f"\t train log_loss: {log_loss_:.4f}")
                eval_df['log_loss']=[log_loss_]
                train_results_df=pd.concat([pd.DataFrame({'y_true':y_true}),pd.DataFrame({'y_prob':y_prob})],axis=1)
                train_results_df['train_counter'] =train_counter             
        train_results= pd.concat([train_results,train_results_df],ignore_index=True)        
    else:
        test_counter=test_counter+1
        for m in metrics:
            eval_df['counter']= test_counter
            eval_df['train']= [train]
            test_results_df =pd.concat([pd.DataFrame({'y_true':y_true}),pd.DataFrame({'y_prob':y_prob})],axis=1)
            test_results_df['test_counter'] =test_counter

            if m in ["log_loss", "loss"]:
                log_loss_ = log_loss(y_true, y_prob, eps=1e-7)
                print(f"\t eval log_loss: {log_loss_:.4f}")
                eval_df['log_loss_']=[log_loss_]
                
            elif m == "balanced_accuracy":
                y_pred = np.round(y_prob)
                accuracy = balanced_accuracy_score(y_true, y_pred)
                print(f"\t eval balanced accuracy: {accuracy:.4f}")
                eval_df['accuracy']=[round(accuracy,4)]
                
                
            elif m == "roc_auc":
                roc_auc = roc_auc_score(y_true, y_prob)
                print(f"\t eval roc_auc: {roc_auc:.4f}")
                eval_df['roc_auc']=[round(roc_auc,4)]

            elif m == "pr_auc":
                precision, recall, _ = precision_recall_curve(y_true, y_prob)
                pr_auc = auc(recall, precision)
                eval_df['pr_auc']=[round(pr_auc,4)]
                
            elif m == "precision":
                precision_all = precision_at_k(y_true_list, y_reco_list,
                                               users, k)
                print(f"\t eval precision@{k}: {precision_all:.4f}")
                eval_df['precision_all']=[round(precision_all,4)]
                
            elif m == "recall":
                recall_all = recall_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval recall@{k}: {recall_all:.4f}")
                eval_df['recall_all']=[round(recall_all,4)]
                
            elif m == "map":
                map_all = map_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval map@{k}: {map_all:.4f}")
                eval_df['map_all']=[round(map_all,4)]
                
            elif m == "ndcg":
                ndcg_all = ndcg_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval ndcg@{k}: {ndcg_all:.4f}")
                eval_df['ndcg_all']=[round(ndcg_all,4)]
                
        #print(y_true_list)
        
        df = pd.concat([df,eval_df],ignore_index=True)
        test_results = pd.concat([test_results,test_results_df],ignore_index=True)
        
        #print(test_results_df)
        rec_list = pd.DataFrame.from_dict(y_reco_list, orient="index").reset_index()
        rec_list.columns = ['user', 'rec_1','rec_2','rec_3', 'rec_4','rec_5','rec_6', 'rec_7','rec_8','rec_9', 'rec_10']
        true_list = pd.DataFrame.from_dict(y_true_list, orient="index").reset_index()
        true_list.columns = ['user', 'true_1', 'true_2', 'true_3' ,'true_4']
        
        
        #print(true_list) 
        #print(rec_list) 
    return df,train_results,test_results

```

As a result of a smaller dataset I noticed that it may cause a syntax issues in some place. _Just something to be aware of._
```
def compute_preds(self,model, data, batch_size, mode, n_users=None, n_items=None):
    y_pred = list()
    y_label = list()
    
    
    print(len(data))
    print(batch_size)
    for batch_data in tqdm(range(0, len(data), batch_size), desc="eval_pred"):
        batch_slice = slice(batch_data, batch_data + batch_size)
        users = data.user_indices[batch_slice]
        items = data.item_indices[batch_slice]
        labels = data.labels[batch_slice]
                
        if mode == "eval":
            user_allowed = np.where(
                np.logical_and(users >= 0, users < n_users))[0]
            item_allowed = np.where(
                np.logical_and(items >= 0, items < n_items))[0]
            indices = np.intersect1d(user_allowed, item_allowed)
            users = users[indices]
            items = items[indices]
            labels = labels[indices]
        preds = list(model.predict(users, items, inner_id=True))


        y_pred.extend(preds)
        y_label.extend(labels)
        ####edit this into the original code
        
        df_pred = pd.DataFrame({'items':items,'users':users,'labels':labels,'y_pred':y_pred,'y_label':y_label,'mode':mode})
        self.df_batch  = pd.concat([self.df_batch ,df_pred],ignore_index=True)
        print(df_pred)
        print(self.df_batch)
        self.df_batch.to_parquet('C:\\Projects\\RecommendationModel\\LibRecommender-master\\LibRec\\output\\df_pred_eval.par', engine='pyarrow', compression='snappy', index=False)
    return y_pred, y_label
```
The output will be repeated since the len(data) < batch_size which is set to 8192. Therefore, it will repeat the range 0 to 1252.

**Comment by Shadz13** [comment]
I have updated **_compute_pred_**
```
def compute_preds(self,model, data, batch_size, mode, n_users=None, n_items=None):
    y_pred = list()
    y_label = list()
    

    print('The len data: ',len(data))
    print('The batch_size data: ',batch_size)
    for batch_data in tqdm(range(0, len(data), batch_size), desc="eval_pred"):
        batch_slice = slice(batch_data, batch_data + batch_size)
        users = data.user_indices[batch_slice]
        items = data.item_indices[batch_slice]
        labels = data.labels[batch_slice]
        print('The batch_data: ',batch_data)
        print('len batch_slice:',batch_slice)
        print('len data labels:',len(data.labels))               
        if mode == "eval":
            user_allowed = np.where(
                np.logical_and(users >= 0, users < n_users))[0]
            item_allowed = np.where(
                np.logical_and(items >= 0, items < n_items))[0]
            indices = np.intersect1d(user_allowed, item_allowed)
            users = users[indices]
            items = items[indices]
            labels = labels[indices]
        preds = list(model.predict(users, items, inner_id=True))


        y_pred.extend(preds)
        y_label.extend(labels)
        ####edit this into the original code
        print('Confirm length')
        print(len(items))
        print(len(users))
        print(len(labels))
        print(len(preds))
        print(len(labels))
        df_pred = pd.DataFrame({'items':items,'users':users,'labels':labels,'preds':preds,'labels':labels,'mode':mode})
        self.df_batch  = pd.concat([self.df_batch ,df_pred],ignore_index=True)
        print(df_pred)
        print(self.df_batch)
        self.df_batch.to_parquet('C:\\Projects\\RecommendationModel\\LibRecommender-master\\LibRec\\output\\df_pred_eval.par', engine='pyarrow', compression='snappy', index=False)
    return y_pred, y_label
```

which stores the following:
```
       items  users  labels     preds  mode
0         14     26     1.0  0.431868  eval
1         26     26     0.0  0.430492  eval
2         22     26     1.0  0.430585  eval
3          1     26     0.0  0.437066  eval
4         15     26     1.0  0.434301  eval
     ...    ...     ...       ...   ...
12267     13   9962     0.0  0.329693  eval
12268     23   9962     1.0  0.297112  eval
12269     24   9962     0.0  0.278629  eval
12270     22   9962     1.0  0.309049  eval
12271     13   9962     0.0  0.329693  eval
```
This is just part of the information used for evaluation. However, just to note that eval_batch_size should be updated throughout the evaluation script. It has be less than len(data), otherwise it produce the same output x amount of times.

NB: I am currently use a small sample set for testing hence I discovered this, otherwise if you are using large set you probably wont encounter this. The default eval_batch_size is 8192.

**Comment by Shadz13** [comment]
Below is the updated function 
```
def print_metrics_ranking(self,df,metrics, y_prob=None, y_true=None, y_reco_list=None,
                          y_true_list=None, users=None, k=10, train=True):

    eval_df=pd.DataFrame()
    
    if train:
        for m in metrics:
            if m in ["log_loss", "loss"]:
                log_loss_ = log_loss(y_true, y_prob, eps=1e-7)
                print(f"\t train log_loss: {log_loss_:.4f}")
                eval_df['log_loss']=[log_loss_]      
    else:
        for m in metrics:

            if m in ["log_loss", "loss"]:
                log_loss_ = log_loss(y_true, y_prob, eps=1e-7)
                print(f"\t eval log_loss: {log_loss_:.4f}")
                eval_df['log_loss_']=[log_loss_]
                
            elif m == "balanced_accuracy":
                y_pred = np.round(y_prob)
                accuracy = balanced_accuracy_score(y_true, y_pred)
                print(f"\t eval balanced accuracy: {accuracy:.4f}")
                eval_df['accuracy']=[round(accuracy,4)]
                
                
            elif m == "roc_auc":
                roc_auc = roc_auc_score(y_true, y_prob)
                print(f"\t eval roc_auc: {roc_auc:.4f}")
                eval_df['roc_auc']=[round(roc_auc,4)]

            elif m == "pr_auc":
                precision, recall, _ = precision_recall_curve(y_true, y_prob)
                pr_auc = auc(recall, precision)
                eval_df['pr_auc']=[round(pr_auc,4)]
                
            elif m == "precision":
                precision_all = precision_at_k(y_true_list, y_reco_list,
                                               users, k)
                print(f"\t eval precision@{k}: {precision_all:.4f}")
                eval_df['precision_all']=[round(precision_all,4)]
                
            elif m == "recall":
                recall_all = recall_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval recall@{k}: {recall_all:.4f}")
                eval_df['recall_all']=[round(recall_all,4)]
                
            elif m == "map":
                map_all = map_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval map@{k}: {map_all:.4f}")
                eval_df['map_all']=[round(map_all,4)]
                
            elif m == "ndcg":
                ndcg_all = ndcg_at_k(y_true_list, y_reco_list, users, k)
                print(f"\t eval ndcg@{k}: {ndcg_all:.4f}")
                eval_df['ndcg_all']=[round(ndcg_all,4)]
                
        df = pd.concat([df,eval_df],ignore_index=True)
        
        rec_list = pd.DataFrame.from_dict(y_reco_list, orient="index").reset_index()
        rec_list.columns = ['user', 'rec_1','rec_2','rec_3', 'rec_4','rec_5','rec_6', 'rec_7','rec_8','rec_9', 'rec_10']
        true_list = pd.DataFrame.from_dict(y_true_list, orient="index").reset_index()
        true_list.columns = ['user', 'true_1', 'true_2', 'true_3' ,'true_4', 'true_5', 'true_6', 'true_7' ,'true_8', 'true_9', 'true_10', 'true_11']
        self.rec_results  = pd.concat([self.rec_results ,rec_list],ignore_index=True)
        self.true_results  = pd.concat([self.true_results ,true_list],ignore_index=True)
       
    return df,self.true_results,self.rec_results
```

**Comment by massquantity** [comment]
Sorry I'm not gonna answer these questions because I find it difficult to explain them well. In general these functions are only used for evaluation inside the library, so they can't fulfill your needs.

I plan to release a new version next week, which can allow users to predict and recommend directly from the original data as well as change user or item features. I've finished most of the work, so just tell me what you need and I'll see if I can do some further modification.

It would also be helpful if you show some DataFrames of your desirable outcome.

**Comment by Shadz13** [comment]
Thanks for the feedback. I will definitely test out the new version and looking forward to it. 

The question previously was quiet long as it focused on understanding the logic of the outputs, therefore I have updated the question to one specific aspect to determine if the y_true output is valid in terms of having duplicates and multiple items?

**Comment by massquantity** [comment]
y_true_list is just all of the items that an user consumed in the training data, so it is natural to contain duplicate items.

**Comment by massquantity** [comment]
I've released version `0.6.0`. See the updated [User Guide](https://github.com/massquantity/LibRecommender/tree/master/examples).
